lib/vnet/daemon/service_darwin.h

```diff
@@ -46,6 +46,10 @@ typedef struct VnetConfigResult {
   const char *ipv6_prefix;
   const char *dns_addr;
   const char *home_path;
+  // egid is the effective group ID of the process on the other side of the XPC connection.
+  gid_t egid;
+  // euid is the effective user ID of the process on the other side of the XPC connection.
+  uid_t euid;
```

Comment:
> Do we have to be concerned about the size of `struct VnetConfigResult` changing and two processes potentially disagreeing about it (an older service running and a freshly upgraded client connecting to it) or is it fine as long as the new fields are at the end of the struct?

The struct we have to worry about is `VnetConfig` from `common_darwin.h`, because that's the one that is sent between processes. That aside, I've not thought about it. [The docs](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingXPCServices.html#//apple_ref/doc/uid/10000172i-SW6-SW11) mention that you can send:

> *   Arithmetic types (`int`, `char`, `float`, `double`, `uint64_t`, `NSUInteger`, and so on)
> *   `BOOL`
> *   C strings
> *   C structures and arrays containing only the types listed above
> *   Objective-C objects that implement the [NSSecureCoding](https://developer.apple.com/documentation/foundation/nssecurecoding) protocol.

So I happily assumed that I don't have to worry about serializing those structs myself. But in hindsight, there's little chance that this seamlessly handles backwards and forwards compatibility.

I just gave it a spin and adding a new bool field at the front or another string field at the end results in an exception being raised and thus the XPC connection being dropped. So XPC seems to be smart enough to at least figure this part and not let the incompatible message through. A connection being dropped is gracefully handled by the client and the daemon, but the actual exception is visible only in Console.app – I believe I'd need to catch it to include it in our logs. Reordering existing string fields on the client results in values being assigned to wrong fields on the daemon side.

<details>
<summary>Exceptions</summary>

Adding a bool at the front (`VnetConfig=****` vs `VnetConfig=B****`):

> `<NSXPCConnection: 0x600002ce43c0> connection from pid 2949 on mach service named com.goteleport.tshdev.vnetd: Exception caught during decoding of received selector startVnet:completion:, dropping incoming message.`
> `Exception: <NSXPCDecoder: 0x14b020400> received a message (startVnet:completion:) with an incompatible message signature (wire: v32@0:8^{VnetConfig=B****}16@?24 vs local: v32@0:8^{VnetConfig=****}16@?24)`

Adding another string field (`****` vs `*****`):

> `<NSXPCConnection: 0x600003cdc280> connection from pid 3023 on mach service named com.goteleport.tshdev.vnetd: Exception caught during decoding of received selector startVnet:completion:, dropping incoming message.`
> `Exception: <NSXPCDecoder: 0x117814000> received a message (startVnet:completion:) with an incompatible message signature (wire: v32@0:8^{VnetConfig=*****}16@?24 vs local: v32@0:8^{VnetConfig=****}16@?24)`

</details>

This is not great, because differences in versions between the client and the daemon can happen. In a separate PR, I'll try to refactor this to a custom Obj-C object with `NSSecureCoding`. It's going to add a bunch of extra boilerplate, but I believe it's going to provide similar compatibility guarantees to protos and much better error messages.

Reply:
For now I'd just leave a comment near `VnetConfig` warning against changing the struct except by appending fields (which will change the size of the message and raise an exception from the XPC side of things). 👍